query-algebrizer/src/clauses/or.rs

```diff
@@ -78,3 +78,289 @@ fn _simply_matches_value_place(left: &PatternValuePlace, right: &PatternValuePla
         _ => true,
     }
 }
+
+pub enum DeconstructedOrJoin {
+    KnownSuccess,
+    KnownEmpty(EmptyBecause),
+    Unit(OrWhereClause),
+    UnitPattern(Pattern),
+    Simple(Vec<Pattern>),
+    Complex(OrJoin),
+}
+
+/// Application of `or`. Note that this is recursive!
+impl ConjoiningClauses {
+    fn apply_or_join(&mut self, schema: &Schema, mut or_join: OrJoin) -> Result<()> {
+        // Simple optimization. Empty `or` clauses disappear. Unit `or` clauses
+        // are equivalent to just the inner clause.
+        match or_join.clauses.len() {
+            0 => Ok(()),
+            1 => match or_join.clauses.pop().unwrap() {
+                OrWhereClause::Clause(clause) => self.apply_clause(schema, clause),
+
+                // A query might be:
+                // [:find ?x :where (or (and [?x _ 5] [?x :foo/bar 7]))]
+                // which is equivalent to dropping the `or` _and_ the `and`!
+                OrWhereClause::And(clauses) => {
+                    for clause in clauses {
+                        self.apply_clause(schema, clause)?;
+                    }
+                    Ok(())
+                },
+            },
+            _ => self.apply_non_trivial_or_join(schema, or_join),
+        }
+    }
+
+    /// Find out if the `OrJoin` is simple. A simple `or` is one in
+    /// which:
+    /// - Every arm is a pattern, so that we can use a single table alias for all.
+    /// - Each pattern should run against the same table, for the same reason.
+    /// - Each pattern uses the same variables. (That's checked by validation.)
+    /// - Each pattern has the same shape, so we can extract bindings from the same columns
+    ///   regardless of which clause matched.
+    ///
+    /// Like this:
+    ///
+    /// ```
+    /// [:find ?x
+    ///  :where (or [?x :foo/knows "John"]
+    ///             [?x :foo/parent "Ámbar"]
+    ///             [?x :foo/knows "Daphne"])]
+    /// ```
+    ///
+    /// While we're doing this diagnosis, we'll also find out if:
+    /// - No patterns can match: the enclosing CC is known-empty.
+    /// - Some patterns can't match: they are discarded.
+    /// - Only one pattern can match: the `or` can be simplified away.
+    fn deconstruct_or_join(&self, schema: &Schema, mut or_join: OrJoin) -> DeconstructedOrJoin {
+        // If we have explicit non-maximal unify-vars, we *can't* simply run this as a
+        // single pattern --
+        // ```
+        // [:find ?x :where [?x :foo/bar ?y] (or-join [?x] [?x :foo/baz ?y])]
+        // ```
+        // is *not* equivalent to
+        // ```
+        // [:find ?x :where [?x :foo/bar ?y] [?x :foo/baz ?y]]
```

Comment:
It would be equivalent to `[:find ?x :where [?x :foo/bar ?y] [?x :foo/baz (gen-sym)]]`, right?  Is there value in deconstructing `or-join` clauses in this way generally?  (Perhaps you do this -- I will see.)